### PR TITLE
validation based check on generate_form_field method

### LINKED ISSUE #3016: validation based check on generate_form_field method
### Describe the problem

No validation on shown `datetimepicker`

### Describe the solution you'd like

Add a case for past date and future Date validation in `generate_form_field`

### PR DESCRIPTION
**Pull request by im-Amitto**
Fixes #3016 

#### Short description of what this resolves:
Added validation for past_date and future_date

I'm not completely sure about the `seq` field in `database.sql`

below added by bradymiller:
Up For Grabs demo for this PR is here:
https://www.open-emr.org/wiki/index.php/Development_Demo#Eta_-_Up_For_Grabs_Demo

### CODE REVIEW

File: interface/new/new_comprehensive.php [hunk 1]
```diff
@@ -948,6 +948,14 @@ function end_group()
...
     <?php // can add any additional javascript settings to datetimepicker here; need to prepend first setting with a comma ?>
   });
+  $('.datepicker-past').datetimepicker({
+    <?php $datetimepicker_timepicker = false; ?>
+    <?php $datetimepicker_showseconds = false; ?>
+    <?php $datetimepicker_formatInput = true; ?>
+    <?php $datetimepicker_maxDate = '+1970/01/01'; ?>
+    <?php require($GLOBALS['srcdir'] . '/js/xl/jquery-datetimepicker-2-5-4.js.php'); ?>
+    <?php // can add any additional javascript settings to datetimepicker here; need to prepend first setting with a comma ?>
+  });
```
Comment (bradymiller) [inline interface/new/new_comprehensive.php]: hi @im-Amitto ,
This is really nice work overall, but I think this means that each form will need to contain a datepicker-past (and future and for the combination if need to support both with time and without time) which I think results in 4 more classes or I think the datepicker will not work. I'm not sure how to solve this unless @mdsupport has an idea to do that?

File: sql/database.sql [hunk 2]
```diff
@@ -4707,6 +4707,7 @@ INSERT INTO `list_options` ( list_id, option_id, title) VALUES ( 'lists','LBF_Va
 INSERT INTO `list_options` (`list_id`,`option_id`,`title`,`notes`, `seq`) VALUES ('LBF_Validations','int1','Integers1-100','{\"numericality\": {\"onlyInteger\": true,\"greaterThanOrEqualTo\": 1,\"lessThanOrEqualTo\":100}}','10');
 INSERT INTO `list_options` (`list_id`,`option_id`,`title`,`notes`, `seq`) VALUES ('LBF_Validations','names','Names','{"format\":{\"pattern\":\"[a-zA-z]+([ \'-\\\\s][a-zA-Z]+)*\"}}','20');
 INSERT INTO `list_options` (`list_id`,`option_id`,`title`,`notes`, `seq`) VALUES ('LBF_Validations','past_date','Past Date','{\"pastDate\":{\"message\":\"must be past date\"}}','30');
+INSERT INTO `list_options` (`list_id`,`option_id`,`title`,`notes`, `seq`) VALUES ('LBF_Validations','future_date','Future Date','{\"futureDate\":{\"message\":\"must be future date\"}}','32');
```
Comment (bradymiller) [inline sql/database.sql]: Regarding seq, this is what is used to order the list when display in the selector. We generally increment by 10 when initially creating a list to allow flexibility to place it in proper order without needing seq of other items. You did the right thing here by making it the number in the middle of the items where you want it do display on the selector (and now still easy to add other items either directly below or above your item in the future).

File: interface/new/new_comprehensive.php (hunk 1 above)
Comment (bradymiller) [reply interface/new/new_comprehensive.php]: Upon thinking about this, the layouts date picker is only used in a finite number of places (maybe 6 or so), so if did need to hard-code the 4 classes it definitely wouldn't be the end of the world to need to add these:
datepicker-past
datetimepicker-past
datepicker-future
datetimepicker-future

**Comment by bradymiller** [comment]
when you test, ensure you try it using the different options in the following 2 settings in Administration->Globals (it should work, but never know):
https://www.open-emr.org/wiki/index.php/Administration_Globals#Date_Display_Format
https://www.open-emr.org/wiki/index.php/Administration_Globals#Time_Display_Format

**Comment by bradymiller** [comment]
I added a Up For Grabs demo for this PR here:
https://www.open-emr.org/wiki/index.php/Development_Demo#Eta_-_Up_For_Grabs_Demo
(this is a openemr demo that runs from this PR; it updates once daily from most recent code in this PR; makes it easy for folks to test it; it's part of [OpenEMR's demo farm](https://github.com/openemr/demo_farm_openemr); if you ever need one of these on a branch you are working on in github, just let me know)

**Comment by bradymiller** [comment]
review complete

File: interface/new/new_comprehensive.php (hunk 1 above)
Comment (im-Amitto) [reply interface/new/new_comprehensive.php]: by hardcode do you mean that i should add these four classes to all files using them or i am interpreting it wrong.

Comment (bradymiller) [reply interface/new/new_comprehensive.php]: yep, I think that is the only way to support that feature and for them to not break. There will then be 6 classes. After finding all the places you need to put these, I bet we can centralize them into a call anywhere that uses the layouts templating.

Comment (im-Amitto) [reply interface/new/new_comprehensive.php]: datetimepicker is currently present in 130 files.

Comment (tywrenn) [reply interface/new/new_comprehensive.php]: I might be totally wrong here but why are we not making certain variables optional in here? We could technically code it with object-oriented programming techniques...

**Comment by mdsupport** [comment]
@im-Amitto 
Short term datepicker specific solution for optional specification of min/max date is what you are doing except - 
I would add empty/isset check in xl/js library.
After its use in the library, I would unset the variable essentially forcing devs to specify these fields if needed for next binding. 

Other possibility (my preference) that you can explore is 
1. Pass data-* at the original field in html part of the script.  I personally prefer since the code becomes much clearer at source.
2. Change the javascript code emitted by xl/js library to change the object being passed on to the control.  Something like : 
```
<input class="datepicker" data-maxdate="now">
...
$("input.datepicker[data-maxdate]").each( ...

```
If you do not want to change existing setup of datetimepicker, you also have an option to just add that property to datetimepicker after the current js binding in each source using the statement.  This will duplicate code but give flexibility for local actions.

Longer term solution would be to 
1. Create oe-ui datetime class that php can set properties in model part of the script.
2. Provide render method(s) that will return appropriate fields that rely on (my personal preference) html5 date and times to take care of calendar and time controls as provided by browsers.  OR if datepicker or any other component is preferred, that code will be centrally generated making all php scripts safe from any changes.
3. That class will also return js code for client side date time validation (min/max and more) that manipulate BS form validation settings.

**Comment by im-Amitto** [comment]
@mdsupport both ideas are pretty great but we can go only with one so "I choose you LTS".

**Comment by im-Amitto** [comment]
@bradymiller should i try the HTML5 datepicker thing in this PR itself or we can change it later?

**Comment by bradymiller** [comment]
hi @im-Amitto , We could definitely bring this in now (I would just do some testing first and would demonstrate our crazy way of supporting mysql changes for upgrade). And then could attempt to migrate these elements to html5 to see if it will work out. Just let me know the route you want to go. thanks!

**Comment by im-Amitto** [comment]
@bradymiller Let's add this quickfix/feature for now and i will start working on html5 datepicker sometime soon.

**Comment by im-Amitto** [comment]
@bradymiller This branch is now 50+ commits behind so rebasing it :smile:

**Comment by stephenwaite** [comment]
nice work @im-Amitto , really impressed with the impact you've had in such a short :timer_clock: 
think @bradymiller was getting at the sql_upgrade script in a prior post
so you'll want to add to the https://github.com/openemr/openemr/blob/master/sql/5_0_2-to-5_0_3_upgrade.sql
like the below
https://github.com/openemr/openemr/blob/master/sql/4_2_2-to-5_0_0_upgrade.sql#L624

**Comment by im-Amitto** [comment]
Thanks @stephenwaite but i can't ignore the fact that i get a lot of help and feedback from @bradymiller, @tywrenn, @mdsupport, @sjpadgett, @thesmallstar, @stu01509, you and the community.
Sometime i even got solutions :see_no_evil:   

Also, I am not sure if this is correct way of doing it?
Screenshot from 2020-03-07 02-25-08

**Comment by stephenwaite** [comment]
hi @im-Amitto , of course, lotta :muscle: involved :)
you'll want to add just your change in database.sql
```
 INSERT INTO `list_options` (`list_id`,`option_id`,`title`,`notes`, `seq`) VALUES ('LBF_Validations','future_date','Future Date','{\"futureDate\":{\"message\":\"must be future date\"}}','32');
```
to the 5.03 upgrade script at the end but before the `#Endif`

**Comment by stephenwaite** [comment]
with this condition on top of course
`#IfNotRow2D list_options list_id lists option_id LBF_Validations`

**Comment by bradymiller** [comment]
hi @im-Amitto , sorry for the delay on this. Planning to review/test this evening after a couple more cups of coffee :)

File: sql/5_0_2-to-5_0_3_upgrade.sql [hunk 3]
```diff
@@ -416,3 +416,7 @@ CREATE TABLE `pro_assessments` (
   PRIMARY KEY (`id`)
 ) ENGINE=InnoDB AUTO_INCREMENT=1;
 #EndIf
+
+#IfNotRow2D list_options list_id lists option_id LBF_Validations
```
Comment (bradymiller) [inline sql/5_0_2-to-5_0_3_upgrade.sql]: Hi @im-Amitto , This PR is testing nicely and bringing into codebase. Very nice work! I am going to make a change to the above line in a subsequent commit, which should be:
```
#IfNotRow2D list_options list_id LBF_Validations option_id future_date
```
Note this "function" will then only run the sql statement here if the there is not yet an entry in list_options with list_id LBF_Validations and option_id future_date

To see all the possible "functions" check out the top of this script.

Comment (bradymiller) [reply sql/5_0_2-to-5_0_3_upgrade.sql]: @im-Amitto ,
Here's the subsequent commit with code change. Also note that we increment $v_database in version.php whenever we make a database change in upgrade script. This is so the brave souls that run OpenEMR of the development codebase know when they need to run upgrade script on the database (using the admin.php script).
https://github.com/openemr/openemr/commit/0be0d4eae37366ce4cba69f16cb86917a35fd182

**Pull request merged**